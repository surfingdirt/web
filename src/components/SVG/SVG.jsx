import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import SVGInline from 'react-svg-inline';

export default class SVG extends PureComponent {
  static propTypes = {
    // Whether to draw the icon as already activated.
    active: PropTypes.bool,
    className: PropTypes.string,
    // Hollow is for vectors where the outline is rendered through 'fill', not 'stroke'.
    hollow: PropTypes.bool,
    // The actual SVG string
    icon: PropTypes.string.isRequired,
    // The label to associate with the icon
    label: (props, propName) => {
      if (props[propName] === '') {
        throw new Error('Empty label should be replaced with presentationOnly.');
      }

      if (
        typeof props.onClick === 'function' &&
        typeof props[propName] === 'undefined' &&
        !props.presentationOnly
      ) {
        throw new Error('label is mandatory when providing an onClick listener.');
      }
      if (!props.presentationOnly && typeof props[propName] === 'undefined') {
        throw new Error('label is mandatory when icon is not just for presentation purposes.');
      }
    },
    // Click listener function
    onClick: PropTypes.func,
    // True if the icon is just for looks and should not be visible to screen readers.
    presentationOnly: PropTypes.bool,
    // Apply standard styles
    standardIcon: PropTypes.bool,
  };

  static defaultProps = {
    active: false,
    className: null,
    hollow: false,
    label: null,
    onClick: null,
    presentationOnly: false,
    standardIcon: false,
  };

  render() {
    const { className, active, icon, label, hollow, onClick, standardIcon } = this.props;

    const actualClassName = classnames(
      className,
      active ? 'active' : '',
      hollow ? 'hollow' : 'full',
      standardIcon ? 'standardIcon' : '',
    );

    const attributes = { svg: icon, className: actualClassName };
    if (label) {
      attributes['aria-label'] = label;
    }
    if (onClick) {
      attributes.onClick = onClick;
      attributes.role = 'button';
      attributes['aria-pressed'] = active;
    }

    return <SVGInline {...attributes} />;
  }
}
