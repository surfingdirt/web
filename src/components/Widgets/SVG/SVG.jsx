import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import SVGInline from 'react-svg-inline';

import { sizes } from 'Utils/icons';
const { SMALL, STANDARD, TINY } = sizes;

export default class SVG extends PureComponent {
  static propTypes = {
    // Whether to draw the icon as already activated.
    active: PropTypes.bool,
    className: PropTypes.string,
    // Hollow is for vectors where the outline is rendered through 'fill', not 'stroke'.
    hollow: PropTypes.bool,
    // The actual SVG string
    icon: PropTypes.string.isRequired,
    // The label to associate with the icon
    label: (props, propName) => {
      if (props[propName] === '') {
        throw new Error('Empty label should be replaced with presentationOnly.');
      }

      if (
        typeof props.onClick === 'function' &&
        typeof props[propName] === 'undefined' &&
        !props.presentationOnly
      ) {
        throw new Error('label is mandatory when providing an onClick listener.');
      }
      if (!props.presentationOnly && typeof props[propName] === 'undefined') {
        throw new Error('label is mandatory when icon is not just for presentation purposes.');
      }
    },
    // Click listener function
    onClick: PropTypes.func,
    // True if the icon is just for looks and should not be visible to screen readers.
    presentationOnly: PropTypes.bool,
    size: PropTypes.string,
  };

  static defaultProps = {
    active: false,
    className: null,
    hollow: false,
    label: null,
    onClick: null,
    presentationOnly: false,
    size: null,
  };

  render() {
    const { className, active, icon, label, hollow, onClick, size } = this.props;

    const actualClassName = classnames(className, hollow ? 'hollow' : 'full', {
      active,
      standardIcon: size === STANDARD,
      smallIcon: size === SMALL,
      tinyIcon: size === TINY,
    });

    const attributes = { svg: icon, className: actualClassName };
    if (label) {
      attributes['aria-label'] = label;
    }
    if (onClick) {
      attributes.onClick = onClick;
      attributes.role = 'button';
      attributes['aria-pressed'] = active;
    }

    return <SVGInline {...attributes} />;
  }
}
