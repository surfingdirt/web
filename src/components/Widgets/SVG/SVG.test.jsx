import React from 'react';
import SVG from './SVG';

// SVG objects passed in are actually strings!
const Icon = `<svg><g></g></svg>`;

describe('SVG', () => {
  it('should render a full icon', () => {
    const wrapper = mount(<SVG icon={Icon} presentationOnly />);
    expect(wrapper).toMatchSnapshot();

    const svg = wrapper.find('span.SVGInline');
    expect(svg).toHaveLength(1);

    const classNames = svg.props().className;
    expect(classNames).toContain('full');
    expect(classNames).not.toContain('hollow');
    expect(classNames).not.toContain('active');
  });

  it('should render a hollow icon', () => {
    const wrapper = mount(<SVG icon={Icon} hollow presentationOnly />);
    expect(wrapper).toMatchSnapshot();

    const svg = wrapper.find('span.SVGInline');
    expect(svg).toHaveLength(1);

    const classNames = svg.props().className;
    expect(classNames).not.toContain('full');
    expect(classNames).toContain('hollow');
    expect(classNames).not.toContain('active');
  });

  it('should render an active icon', () => {
    const wrapper = mount(<SVG icon={Icon} active presentationOnly />);
    expect(wrapper).toMatchSnapshot();

    const svg = wrapper.find('span.SVGInline');
    expect(svg).toHaveLength(1);

    const classNames = svg.props().className;
    expect(classNames).toContain('full');
    expect(classNames).not.toContain('hollow');
    expect(classNames).toContain('active');
  });

  it('should render an active hollow icon', () => {
    const wrapper = mount(<SVG icon={Icon} hollow active presentationOnly />);
    expect(wrapper).toMatchSnapshot();

    const svg = wrapper.find('span.SVGInline');
    expect(svg).toHaveLength(1);

    const classNames = svg.props().className;
    expect(classNames).not.toContain('full');
    expect(classNames).toContain('hollow');
    expect(classNames).toContain('active');
  });

  it('should add a class name', () => {
    const wrapper = mount(<SVG icon={Icon} className="toto" presentationOnly />);
    expect(wrapper).toMatchSnapshot();

    const svg = wrapper.find('span.SVGInline');
    expect(svg).toHaveLength(1);

    const classNames = svg.props().className;
    expect(classNames).toContain('toto');
  });

  it('should not render a label', () => {
    const wrapper = mount(<SVG icon={Icon} label="my icon" />);
    expect(wrapper).toMatchSnapshot();

    const svg = wrapper.find('span.SVGInline');
    expect(svg).toHaveLength(1);

    expect(svg.props()['aria-label']).toEqual('my icon');
  });
});
