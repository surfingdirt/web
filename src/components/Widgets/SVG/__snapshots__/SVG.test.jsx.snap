// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVG should add a class name 1`] = `
<span
  className="SVGInline toto full"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"SVGInline-svg toto-svg full-svg\\"><g></g></svg>",
    }
  }
/>
`;

exports[`SVG should not render a label 1`] = `
<span
  aria-label="my icon"
  className="SVGInline full"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"SVGInline-svg full-svg\\"><g></g></svg>",
    }
  }
/>
`;

exports[`SVG should render a full icon 1`] = `
<span
  className="SVGInline full"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"SVGInline-svg full-svg\\"><g></g></svg>",
    }
  }
/>
`;

exports[`SVG should render a hollow icon 1`] = `
<span
  className="SVGInline hollow"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"SVGInline-svg hollow-svg\\"><g></g></svg>",
    }
  }
/>
`;

exports[`SVG should render an active hollow icon 1`] = `
<span
  className="SVGInline active hollow"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"SVGInline-svg active-svg hollow-svg\\"><g></g></svg>",
    }
  }
/>
`;

exports[`SVG should render an active icon 1`] = `
<span
  className="SVGInline active full"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<svg class=\\"SVGInline-svg active-svg full-svg\\"><g></g></svg>",
    }
  }
/>
`;
